numpy/core/src/multiarray/compiled_base.c

```diff
@@ -1120,6 +1120,30 @@ arr_unravel_index(PyObject *self, PyObject *args, PyObject *kwds)
         goto fail;
     }
 
+    if (!PyArray_CanCastArrayTo(indices, dtype, NPY_SAFE_CASTING)) {
+        if (PyArray_SIZE(indices) == 0) {
+            /*
+             * Unfortunately, if the input array is empty, then indices would
+             * likely have a default dtype of float64, which isn't compatibile.
+             * Force a recast of an empty array.
+             */
+            PyArrayObject *indices_tmp = NULL;
+            Py_INCREF(dtype);
+            indices_tmp = (PyArrayObject*)PyArray_FromArray(indices, dtype,
+                                                            NPY_ARRAY_FORCECAST);
+            if (indices_tmp == NULL) {
+                goto fail;
+            }
+            Py_DECREF(indices);
+            indices = indices_tmp;
+        }
+        else {
+            PyErr_SetString(PyExc_ValueError,
```

Comment:
Just wondering, was the iterator error message so much worse? (I guess it is a bit overly technical in comparison).